consensus/contracts.go

```diff
@@ -36,274 +36,144 @@ func (s *State) storageProofSegmentIndex(contractID ContractID, windowIndex Bloc
-				}
-
-				// Create the output.
-				outputID := ContractTerminationOutputID(openContract.ContractID, contractSuccess)
-				s.unspentOutputs[outputID] = output
-				diff := OutputDiff{New: true, ID: outputID, Output: output}
-				diffs = append(diffs, diff)
-			}
-
-			// Add the contract to contract terminations.
-			s.currentBlockNode().ContractTerminations = append(s.currentBlockNode().ContractTerminations, openContract)
+	// Update the state.
+	s.unspentOutputs[outputID] = output
+	delete(s.openContracts, id)
 
-			// Mark contract for deletion (can't delete from a map while
-			// iterating through it - results in undefined behavior of the
-			// iterator.
-			contractsToDelete = append(contractsToDelete, openContract.ContractID)
-		}
+	cd = ContractDiff{
+		New:      false,
+		ID:       id,
+		Contract: contract,
 	}
-
-	// Delete all of the contracts that terminated.
-	for _, contractID := range contractsToDelete {
-		delete(s.openContracts, contractID)
+	od = OutputDiff{
+		New:    true,
+		ID:     outputID,
+		Output: output,
 	}
 	return
 }
 
-// inverseContractMaintenance does the inverse of contract maintenance, moving
-// the state of contracts backwards instead forwards.
-func (s *State) invertContractMaintenance() (diffs []OutputDiff) {
-	// Repen all contracts that terminated, and remove the corresponding output.
-	for _, openContract := range s.currentBlockNode().ContractTerminations {
-		id := openContract.ContractID
-		s.openContracts[id] = openContract
-		contractStatus := openContract.Failures == openContract.FileContract.Tolerance
-		outputID := ContractTerminationOutputID(id, contractStatus)
-		diff := OutputDiff{New: false, ID: outputID, Output: s.unspentOutputs[outputID]}
-		delete(s.unspentOutputs, outputID)
-		diffs = append(diffs, diff)
+func (s *State) applyContractMaintenance() (outputDiffs []OutputDiff, contractDiffs []ContractDiff) {
+	// Iterate through all contracts and figure out which ones have expired.
+	// Expiring a contract deletes it from the map we are iterating through, so
+	// we need to store it and deleted once we're done iterating through the
+	// map.
+	//
+	// TODO: iterating through everything is inefficient, but sufficient for
+	// now.
+	var expiredContracts []ContractID
+	for id, contract := range s.openContracts {
+		if s.height() == contract.End {
+			expiredContracts = append(expiredContracts, id)
+		}
 	}
 
-	// Reverse all outputs created by missed storage proofs.
-	for _, missedProof := range s.currentBlockNode().MissedStorageProofs {
-		cid, oid := missedProof.ContractID, missedProof.OutputID
-		s.openContracts[cid].FundsRemaining += s.unspentOutputs[oid].Value
-		s.openContracts[cid].Failures -= 1
-		diff := OutputDiff{New: false, ID: oid, Output: s.unspentOutputs[oid]}
-		delete(s.unspentOutputs, oid)
-		diffs = append(diffs, diff)
+	// Delete all of the contracts that terminated.
+	for _, id := range expiredContracts {
+		contract := s.openContracts[id]
```

Comment:
if you're just going to lookup the contract again, why not make `expiredContracts` a `[]*OpenContract` instead of a `ContractID`?